Allow running a single test in testsuite/Makefile

dra27 opened the pull request:
This PR adds a `just` target to `testsuite/Makefile` which is like the `one` target but for a single test.

Story:
- Some tests are just slow; many tests are slow on Windows
- When we had `ocamltests`, I'd occasionally hack those in anger, thinking "there should be a better way of doing this" and use `make one DIR=`
- That better way was sort of `../ocamltest/ocamltest tests/path/to/test.ml`
- But that doesn't work if you bootstrapped FLEXLINK, etc.
- So this PR allows calling ocamltest with the environment set-up properly 🙂

dra27 committed:
Allow running a single test in testsuite/Makefile

shindere commented:
I like the idea, thanks.

The only thing is that I tned to find the names of the make targets a
bit too ad hoc. It was already the case with `one`, so it's not only
about this PR, but IMHO `just` makes things worse. Perhaps we could take
the opportunity of this PR to go for slightly more meaningful names,
like `testdir` and `testfile` or just `test` for the later one?

dra27 commented:
Every time recently I have to type `make PLEASE_KEEP_THE_OCAMLTEST_DIRECTORY_EVEN_ON_SUCCESS=1 ...` I confess I become more 1970s-like in my desire for short names 🙂

Seriously, though - I've never never seen a problem with the `one` target - `make one` is an error, and `make one DIR=directory` is pretty clear. Maybe it could have been called `make only`...

Running temporarily with that idea, perhaps we could merge `one`, `list` and my proposed `just` into a single target where the effect is controlled by the variable defined (`one` and `list` obviously left for human-compatibility)? So `make only DIR=directory` or `make only LIST=list-file` or `make only TEST=test`? That's not actually much code, given that `$(words $(TEST) $(DIR) $(LIST))` gives a trivial way of detecting if you cheekily try to say `make only TEST=hah LIST=tricked DIR=you`

shindere commented:
Well actually `make one` with the variable specifying what exactly to
make would be okay, wouldn't it?

dra27 committed:
Always read actual lines from files

read should, as a general rule, always be followed by -r in scripts. If
IFS is not empty, then spaces are stripped as well.

dra27 committed:
Augment make -C testsuite one with FILE and LIST

dra27 commented:
While looking at this, I noticed that our `while` loops are using `read` badly - I've pushed one commit which changes `while read LINE` to `while IFS='' read -r LINE`. The `-r` stops `read` from doing horrific and non-portable things to backslashes and the `IFS=''` stops it from mangling spaces in the lines. Basically, `while IFS='' read -r LINE` actually reads a line from a file! For more details of at least part of that, see https://github.com/koalaman/shellcheck/wiki/SC2162.

The second commit I've pushed gets rid of the `just` target and allows you to say `make one DIR=tests/dir` (as before), `make one TEST=tests/dir/test.ml` (so what was `make just` at this PR) and also adds `make one LIST=list` which is the same as `make list FILE=list`. The help documentation no longer mentions the `list` target, but it's still there to allow muscle-memory to be retained!

dra27 commented:
Rebased to remove the conflict - how does this look, @shindere?

shindere commented:
David Allsopp (2020/07/31 01:23 -0700):
> Rebased to remove the conflict - how does this look, @shindere?

How about renaming LIST to FILES?

gasche commented:
I like the new interface with DIR and TEST. (no opinion on LIST vs. FILES)

xavierleroy commented:
Can we merge promptly, or is there work remaining?

dra27 commented:
> > Rebased to remove the conflict - how does this look, @shindere?
>
> How about renaming LIST to FILES?

At present, `LIST` is a actually a file with a list of directories, so `FILES` seems a bit strange.

dra27 commented:
@xavierleroy - I got surprised by `make list` only allowing directories and had started extending it to support directories _or_ files (quite useful sometimes where you have a very specific of lists of tests your developing with) but that doesn't have to be done here. I'm happy with this - indeed I missed it recently 🙂

xavierleroy reviewed (approved):
Looks good to me.  I could certainly use `make one FILE=...` myself!

shindere commented:
Fine with me, thanks!
